### PR TITLE
sql: refactor plan{Visitor,Observer} to avoid unnecessary work

### LINKED ISSUE #13421: sql: planVisitor.visit is allocation heavy
`planVisitor.visit` shows up high in allocation profiles. It looks like a little elbow grease could dramatically reduce this. For example, when performing an `INSERT` we call `planner.makePlan` -> `planner.optimizePlan` -> `planVisitor.visit`. The `observer` in this call path is a `subqueryInitializer` which only implements the `planObserver.subqueryNode` method. The other `planObserver` methods are still called even though they aren't used. The unnecessary allocations primarily come from calls to `planObserver.attr`. Specifically, the generation of the `description` for `*valuesNode`:

```go
		suffix := "not yet populated"
		if n.rows != nil {
			suffix = fmt.Sprintf("%d row%s",
				n.rows.Len(), util.Pluralize(int64(n.rows.Len())))
		} else if n.tuples != nil {
			suffix = fmt.Sprintf("%d row%s",
				len(n.tuples), util.Pluralize(int64(len(n.tuples))))
		}
		description := fmt.Sprintf("%d column%s, %s",
			len(n.columns), util.Pluralize(int64(len(n.columns))), suffix)
		v.observer.attr(name, "size", description)
```

One thought for fixing this is to change `planObserver` from an interface to a struct:

```
type planObserver struct {
	// enterNode is invoked upon entering a tree node. It can return false to
	// stop the recursion at this node.
	enterNode func(nodeName string, plan planNode) bool

	// expr is invoked for each expression field in each node.
	expr func(nodeName, fieldName string, n int, expr parser.Expr)

	// attr is invoked for non-expression metadata in each node.
	attr func(nodeName, fieldName, attr string)

	// leaveNode is invoked upon leaving a tree node.
	leaveNode func(nodeName string)

	// subqueryNode is invoked for each sub-query node. It can return
	// an error to stop the recursion entirely.
	subqueryNode func(sq *subquery) error
}
```

Usage would then transform to:

```
	i := &subqueryInitializer{p: p}
        o := planObserver{subqueryNode: i.subqueryNode}
	if err := walkPlan(newPlan, o); err != nil {
		return plan, err
	}
```

I'm going to experiment with this.

### PR DESCRIPTION
**Pull request by petermattis**
Change planObserver from an interface to a struct containing function
pointers. Only generate the data needed for a particular observer
function if that function pointer is non-nil.

Fixes #13421

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13422)

### CODE REVIEW

File: pkg/sql/subquery.go [hunk 1]
```diff
@@ -218,16 +216,15 @@ func (v *subqueryPlanVisitor) enterNode(_ string, n planNode) bool {
...
-func (v *subqueryPlanVisitor) expr(_, _ string, _ int, _ parser.Expr) {}
-func (v *subqueryPlanVisitor) leaveNode(_ string)                     {}
-
 func (p *planner) startSubqueryPlans(plan planNode) error {
 	// We also run and pre-evaluate the subqueries during start,
 	// so as to avoid re-running the sub-query for every row
 	// in the results of the surrounding planNode.
 	p.subqueryPlanVisitor = subqueryPlanVisitor{p: p}
-	return walkPlan(plan, &p.subqueryPlanVisitor)
+	return walkPlan(plan, planObserver{
```
Comment (knz) [inline pkg/sql/subquery.go]: I would suggest providing a default instance where the function pointers are not `nil`, but rather no-op functions, make the users of `walkPlan` override some of them, and drop the checks on `nil` in `visit()` to increase readability.

Comment (petermattis) [inline pkg/sql/subquery.go]: > and drop the checks on nil in visit() to increase readability.

The nil checks in `planVisitor.visit` are the primary reason for this PR. In particular, notice how we don't generate the `attr` field to pass to `planObserver.attr` unless that field is non-nil.

**Comment by knz** [comment]
I think the approach you took here is excessively local.

I'd rather keep `visit` simple and instead change the planObserver interface to *never* require the allocation of objects to pass arguments to it. That would be a less intrusive (and less verbose) change.

I can guide you through it or you can push the PR my way.

Meanwhile I am curious: how significant are these allocation counts relative to other things?

**Comment by petermattis** [comment]
> I'd rather keep visit simple and instead change the planObserver interface to never require the allocation of objects to pass arguments to it. That would be a less intrusive (and less verbose) change.

How would you achieve that?

> I can guide you through it or you can push the PR my way.

I'm fine with iterating on the PR myself. 

> Meanwhile I am curious: how significant are these allocation counts relative to other things?

I don't have the numbers on hand anymore, but I recall single-digit percentage of allocations in an insert-only workload. The unfortunate bit about these allocations and the associated work is that they are completely unnecessary except for `EXPLAIN`.

**Comment by petermattis** [comment]
> I don't have the numbers on hand anymore, but I recall single-digit percentage of allocations in an insert-only workload.

Mentioned to @knz offline: prior to this change `planVisitor.visit` accounted for 9.9% of the allocations in a write-only workload. After this change it accounts for 0.9%.

**Comment by knz** [comment]
I'm sorry I can't find anything better to improve this.
LGTM

**Comment by petermattis** [comment]
> I'm sorry I can't find anything better to improve this.

I'm both relieved and disappointed.

**Pull request merged**